Sources/SwiftProtobuf/Message+AsyncMessageSequence.swift

```diff
@@ -0,0 +1,135 @@
+//
+// Sources/SwiftProtobuf/Message+AsyncMessageSequence.swift - Async sequence over binary delimited protobuf
+//
+// Copyright (c) 2023 Apple Inc. and the project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See LICENSE.txt for license information:
+// https://github.com/apple/swift-protobuf/blob/main/LICENSE.txt
+//
+// -----------------------------------------------------------------------------
+///
+/// Extensions to `Message` to provide an async sequence over a binary delimited protobuf stream.
+///
+// -----------------------------------------------------------------------------
+
+import Foundation
+
+@available(macOS 12.0, iOS 15.0, tvOS 15.0, watchOS 8.0, *)
+extension Message {
+  
+  /// Creates an asynchronous sequence of messages decoded from resource bytes
+  public static func asyncSequence(
```

Comment:
This extension doesn't make too much sense to me. What you rather want is an extension on `AsyncSequence` it self that provides a non-static method like `decodedMessages()` (name can be bike shed)

Reply:
I removed it altogether. TBH I didn't understand what having an extension on AsyncSequence would achieve.

Reply:
Please add this back. The extension is super helpful because it allows you to chain algorithms together, e.g.

```
let someByteSequence = AsyncStream<UInt8>()
someByteSequence.decodeBinaryDelimitedMessages().filter { ...}.map {}
```

Reply:
@FranzBusch I may be missing something here but I feel like the extension on `Message` makes far more sense than an extension on `AsyncStream` .

With the extension on `Message`, originally suggested by @tbkka , the message type is elegantly inferred and is very clear at the call site, e.g: 

```
    let messageExtMethod = SwiftProtoTesting_TestAllTypes
      .asyncSequence(baseSequence: url.resourceBytes).filter { m in
      m.optionalInt32 > 2
    }.map { m in
      m.optionalInt32
    }
    for try await optionalInt in messageExtMethod {
      print(optionalInt)
    }
```

With an extension on AsyncSequence it gets cumbersome and actually makes method chaining more difficult. I haven't found a way to specify the message type other than:

```
    let asyncSeqExtMethod: AsyncMessageSequence<URL.AsyncBytes, SwiftProtoTesting_TestAllTypes>
    = url.resourceBytes.decodedMessages()
    let filteredInt32s = asyncSeqExtMethod.filter { m in
      m.optionalInt32 > 2
    }.map { m in
      m.optionalInt32
    }
    for try await msg in filteredInt32s {
      print(msg)
    }
 ```

(It doesn't seem possible to pass the message type as a parameter).

The `AsyncSequence` extension idea as I understood it:
```
extension AsyncSequence {
  
  public func decodedMessages<M: Message>(
    extensions: ExtensionMap? = nil,
    partial: Bool = false,
    options: BinaryDecodingOptions = BinaryDecodingOptions()
  ) -> AsyncMessageSequence<Self, M> {
    AsyncMessageSequence<Self, M>(
      baseSequence: self,
      extensions: extensions,
      partial: partial,
      options: options
    )
  }
}
```

Reply:
I've put both extensions in there, I guess we can decide which to keep.